lib/Runtime/Library/JSON.cpp

```diff
@@ -647,42 +679,68 @@ namespace JSON
             }
             else
             {
-                uint32 precisePropertyCount = 0;
                 Js::JavascriptStaticEnumerator enumerator;
                 if (object->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EphemeralReference, scriptContext))
                 {
-                    bool isPrecise;
-                    uint32 propertyCount = GetPropertyCount(object, &enumerator, &isPrecise);
-                    if (isPrecise)
-                    {
-                        precisePropertyCount = propertyCount;
-                    }
+                      Js::DynamicObject * dynamicObject = ((Js::DynamicObject*)object);
+                      if (ReplacerFunction != replacerType || !dynamicObject->HasObjectArray())
+                      {
+                        int propertyCount = dynamicObject->GetPropertyCount();
 
-                    if (ReplacerFunction != replacerType)
-                    {
-                        // Reserve initial slots for properties. +1 stands for extra property being pushed during the stringify
-                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount + 1);
+                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
+                        Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
                         enumerator.Reset();
-                        while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
+                        Js::PropertyId previousId;
+                        if (dynamicObject->HasObjectArray())
                         {
-                             if (id == Js::Constants::NoProperty)
+                            int totalNumPropertyCount = this->GetPropertyCount(object, &enumerator) - propertyCount;
+                            Assert(totalNumPropertyCount > 0);
+                            int index = 0;
+                            enumerator.Reset();
+                            while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL && index < totalNumPropertyCount)
                             {
-                                //if unsuccessful get propertyId from the string
+                                // if unsuccessful get propertyId from the string
                                 scriptContext->GetOrAddPropertyRecord(propertyName->GetString(), propertyName->GetLength(), &propRecord);
                                 id = propRecord->GetPropertyId();
+                                StringifyMemberObject(propertyName, id, value, (Js::ConcatStringBuilder*)result, indentString, memberSeparator, isFirstMember, isEmpty);
+                                index++;
+                            }
+                        }
+
+                        for (PropertyIndex i = 0; i < propertyCount; i++)
```

Comment:
I'm a little confused by what's going on here. Could you please add some comments?